### ISSUE TITLE
Replicating 94.0% performance on CIFAR-10

### ISSUE THREAD

**Issue by kibok90** [google-research/simclr#22]
Hi,

I am trying to replicate the numbers reported in Appendix B.9, but stuck around 91~92%.
The training recipe in README gives me around 91% accuracy as stated.

At least increasing depth (ResNet18 -> 50) did not show a meaningful performance gain (~0.5%), and I believe this is consistent with the behavior in the supervised learning on CIFAR-10, which is reported in https://github.com/kuangliu/pytorch-cifar.

If I understand correctly, I should be able to see 93~94% accuracy after 1k epochs with any batch size as reported in Figure B.4, and specifically 94.0% when the batch size is 1024.

Could you share the exact training recipe to replicate the performances?

Thanks.

**Comment by sayakpaul** [comment]
@kibok90 are you using PyTorch? Just asking out of curiosity.

**Comment by kibok90** [comment]
@sayakpaul I tried both PyTorch and TF (this repo).
The best performance in my PyTorch implementation is only around 90% though.

**Comment by sayakpaul** [comment]
Thanks for mentioning @kibok90. I tried a minimal implementation in `tf.keras` with custom loops without leaving out the major pieces. I am yet to get the desired results from the linear evaluation protocol. I am using this dataset: https://github.com/thunderInfy/imagenet-5-categories.

**Comment by chentingpc** [comment]
I think training with ResNet-50, you should be able to get >93% with batch size of 512. I didn't test it here because ResNet-50 is expensive (for 1 GPU) and an overkill for CIFAR-10 (as we directly modify it from ImageNet ResNet-50, so probably a cifar10 wide resnet can work much more efficiently). If you only get 92%, I would suggest: (1) try temperature between 0.1 - 0.5, e.g. 0.2; (2) train the linear evaluation longer.

**Comment by kibok90** [comment]
@chentingpc Thanks for your answer.
Actually, I already tried your suggestions, but couldn't get a performance higher than 92%.

Now I am trying ResNet-50 with the code in this repo. But I confirmed that ResNet-18 -> 50 does not increase the performance significantly in another codebase (I saw only ~0.5% improvement), so I don't expect to achieve 93~94% simply with this model size change though.

So it seems the code in this repo is different from the code used for the experiment in Appendix B.9? Could you double-check if there is any difference between this repo and your code, and/or confirm that this repo can replicate the numbers in Appendix B.9?

**Comment by chentingpc** [comment]
This codebase is consistent with Appendix B.9 (some minor differences due to open source but shouldn't matter much). In my previous experiments on CIFAR10, ResNet-50 is ~2% better than than ResNet-18, so I am not sure how you got only ~0.5% improvement (it is possible that architecture is different, you can refer to resnet.py for our ResNet with cifar stem, which may be different from a your resnet-50 for cifar).

**Comment by sayakpaul** [comment]
I was working on a minimal implementation of SimCLR using CIFAR10 using custom loops and `tf.keras`. Hence I thought of mentioning it here. 

Here's the augmentation policy I am using:

```
# Comes from: https://github.com/google-research/simclr
class CustomAugment(object):
    def __call__(self, sample):
        # Random flips
        sample = self._random_apply(tf.image.flip_left_right, sample, p=0.5)
        
        # randomly apply transformation (color distortions and blur) with probability p.
        sample = self._random_apply(self._color_jitter, sample, p=0.8)
        sample = self._random_apply(self._color_drop, sample, p=0.2)
        sample = self._random_apply(self._gaussian_blur, sample, p=0.5)

        return sample

    def _color_jitter(self, x, s=0.5):
        # one can also shuffle the order of following augmentations
        # each time they are applied.
        x = tf.image.random_brightness(x, max_delta=0.8*s)
        x = tf.image.random_contrast(x, lower=1-0.8*s, upper=1+0.8*s)
        x = tf.image.random_saturation(x, lower=1-0.8*s, upper=1+0.8*s)
        x = tf.image.random_hue(x, max_delta=0.2*s)
        x = tf.clip_by_value(x, 0, 1)
        return x
    
    def _color_drop(self, x):
        image = tf.image.rgb_to_grayscale(x)
        image = tf.tile(x, [1, 1, 1, 3])
        return x

    def _gaussian_blur(self, image, sigma=tf.random.uniform([], 0.1, 2.0, dtype=tf.float32),
                       padding='SAME'):
        kernel_size = image.shape[1]
        radius = tf.cast(kernel_size / 2, dtype=tf.int32)
        kernel_size = radius * 2 + 1
        
        x = tf.cast(tf.range(-radius, radius + 1), tf.float32)
        blur_filter = tf.exp(
          -tf.pow(x, 2.0) / (2.0 * tf.pow(tf.cast(sigma, tf.float32), 2.0)))
        blur_filter /= tf.reduce_sum(blur_filter)
        
        # One vertical and one horizontal filter.
        blur_v = tf.reshape(blur_filter, [kernel_size, 1, 1, 1])
        blur_h = tf.reshape(blur_filter, [1, kernel_size, 1, 1])
        num_channels = tf.shape(image)[-1]
        blur_h = tf.tile(blur_h, [1, 1, num_channels, 1])
        blur_v = tf.tile(blur_v, [1, 1, num_channels, 1])
        
        expand_batch_dim = image.shape.ndims == 3
        if expand_batch_dim:
            # Tensorflow requires batched input to convolutions, which we can fake with
            # an extra dimension.
            image = tf.expand_dims(image, axis=0)
        
        blurred = tf.nn.depthwise_conv2d(
            image, blur_h, strides=[1, 1, 1, 1], padding=padding)
        blurred = tf.nn.depthwise_conv2d(
            blurred, blur_v, strides=[1, 1, 1, 1], padding=padding)
        
        if expand_batch_dim:
            blurred = tf.squeeze(blurred, axis=0)
        
        blurred = tf.clip_by_value(blurred, 0., 1.)
        return blurred
    
    def _random_apply(self, func, x, p):
        return tf.cond(
          tf.less(tf.random.uniform([], minval=0, maxval=1, dtype=tf.float32),
                  tf.cast(p, tf.float32)),
          lambda: func(x),
          lambda: x)
```
Note that I am not using random crops. 

I am using ResNet20 as the base encoder network sourced from [here](https://raw.githubusercontent.com/GoogleCloudPlatform/keras-idiomatic-programmer/master/zoo/resnet/resnet_cifar10.py) from `keras-idiomatic-programmer`. 

I am using an implementation of the NT-XEnt loss from here: https://github.com/google-research/simclr/blob/master/objective.py:

```
def nt_xentloss(hidden_1,
		hidden_2,
		hidden_norm=True,
		temperature=1.0,
		weights=1.0):
  
	# Get (normalized) hidden1 and hidden2.
	if hidden_norm:
		hidden_1 = tf.math.l2_normalize(hidden_1, -1)
		hidden_2 = tf.math.l2_normalize(hidden_2, -1)
	# hidden1, hidden2 = tf.split(hidden, 2, 0)
	batch_size = tf.shape(hidden_1)[0]

	hidden1_large = hidden_1
	hidden2_large = hidden_2
	labels = tf.one_hot(tf.range(batch_size), batch_size * 2)
	masks = tf.one_hot(tf.range(batch_size), batch_size)

	logits_aa = tf.matmul(hidden_1, hidden1_large, transpose_b=True) / temperature
	logits_aa = logits_aa - masks 
	logits_bb = tf.matmul(hidden_2, hidden2_large, transpose_b=True) / temperature
	logits_bb = logits_bb - masks 
	logits_ab = tf.matmul(hidden_1, hidden2_large, transpose_b=True) / temperature
	logits_ba = tf.matmul(hidden_2, hidden1_large, transpose_b=True) / temperature

	loss_a = tf.compat.v1.losses.softmax_cross_entropy(
		labels, tf.concat([logits_ab, logits_aa], 1), weights=weights)
	loss_b = tf.compat.v1.losses.softmax_cross_entropy(
		labels, tf.concat([logits_ba, logits_bb], 1), weights=weights)
	loss = loss_a + loss_b

	return loss, logits_ab, labels
```

Note that I am not multiplying `LARGE_NUM` as it was resulting in `NaN` loss values for my experimental setting. The network seems to have made progress in the training (for 100 epochs):

![image](https://user-images.githubusercontent.com/22957388/81062849-5fe18100-8ef4-11ea-9a7f-7bd6a78de14a.png)

Now, when I am trying linear evaluation with 10% of the training set, here's what I am getting:

![image](https://user-images.githubusercontent.com/22957388/81062927-86072100-8ef4-11ea-8cfb-4351e76c15ad.png)

I tried with different projection heads and no project as well to extract the image embeddings while training the linear model but the performance remained somewhat the same. But I am no way closer to 90% mark yet. I am not using LARS, instead, I am using SGD with a CosineDecay policy. 

Here are the Colab Notebooks for reproducibility:
- [SimCLR training](https://colab.research.google.com/gist/sayakpaul/61be93d0fe7e2d9ab40db45357eac6a2/simclr_in_tensorflow_2_cifar10_idiomatic.ipynb)
- [Linear Evaluation](https://colab.research.google.com/gist/sayakpaul/0902955c9245e3b14427b0f4dfc6739f/linear_evaluation.ipynb) 

Any helping directions would be great.

**Comment by chentingpc** [comment]
@sayakpaul I didn't spot any obvious bug from the code you posted here. If you want to make sure your code reproduces the result in our codebase, you could run both code and make sure the loss curve roughly matches. There may be some small differences that lead to significant difference in result, and it is hard to find out by just looking at this code snippet.

**Comment by kibok90** [comment]
> This codebase is consistent with Appendix B.9 (some minor differences due to open source but shouldn't matter much). In my previous experiments on CIFAR10, ResNet-50 is ~2% better than than ResNet-18, so I am not sure how you got only ~0.5% improvement (it is possible that architecture is different, you can refer to resnet.py for our ResNet with cifar stem, which may be different from a your resnet-50 for cifar).

Thanks for your reply. I ran the experiment with this code in the meantime, and here are the results:
bsz=512, lr=1: 92.65%
bsz=512, lr=1.5: 93.30%

The average performance reported in Figure B.5 is larger than 93.5% so they are still somewhat low, but I think it is acceptable.

**Comment by ClaraBing** [comment]
> I think training with ResNet-50, you should be able to get >93% with batch size of 512. I didn't test it here because ResNet-50 is expensive (for 1 GPU) and an overkill for CIFAR-10 (as we directly modify it from ImageNet ResNet-50, so probably a cifar10 wide resnet can work much more efficiently). If you only get 92%, I would suggest: (1) try temperature between 0.1 - 0.5, e.g. 0.2; (2) train the linear evaluation longer.

May I ask how long it takes to train a ResNet-18 / ResNet-18 on CIFAR10 with a single GPU? Thank you!

**Comment by chentingpc** [comment]
it depends on which GPU, and how long you train it :) I didn't train this with single GPU for a long time, so my vague impression is that it probably should be done within a couple hours (<10) if you train with something like P100 for 500-1000 epochs.

You can also estimate it by double the training time (due to two augmented views) of supervised learning in the same epochs.
